benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
-    max_used -= prv_used;                                          \
-    max_blocks -= prv_blocks;                                      \
-    max_bytes = max_used + MEM_BLOCK_OVERHEAD * max_blocks;        \
-    mbedtls_printf( "%6u heap bytes", (unsigned) max_bytes );
 
-#else
-#define MEMORY_MEASURE_INIT
-#define MEMORY_MEASURE_PRINT( title_len )
-#endif
+static void alarm() {
+    alarmed = 1;
+}
 
-#define TIME_PUBLIC( TITLE, TYPE, CODE )            \
-do {                                                \
-    unsigned long ms;                               \
-    int ret = 0;                                    \
-    Timer t;                                        \
-    MEMORY_MEASURE_INIT;                            \
-                                                    \
-    mbedtls_printf( HEADER_FORMAT, TITLE );         \
-    fflush( stdout );                               \
-                                                    \
-    t.start();                                      \
-    CODE;                                           \
-    t.stop();                                       \
-    ms = t.read_ms();                               \
-                                                    \
-    if( ret != 0 )                                  \
-    {                                               \
-        PRINT_ERROR;                                \
-    }                                               \
-    else                                            \
-    {                                               \
-        mbedtls_printf( "%6lu ms/" TYPE, ms );      \
-        MEMORY_MEASURE_PRINT( sizeof( TYPE ) + 1 ); \
-        mbedtls_printf( "\r\n" );                   \
-    }                                               \
-} while( 0 )
-
-static int myrand( void *rng_state, unsigned char *output, size_t len )
-{
+static int myrand(void *rng_state, unsigned char *output, size_t len) {
     size_t use_len;
     int rnd;
 
-    if( rng_state != NULL )
+    if (rng_state != NULL) {
         rng_state  = NULL;
+    }
 
-    while( len > 0 )
-    {
+    while (len > 0) {
         use_len = len;
-        if( use_len > sizeof(int) )
+        if(use_len > sizeof(int)) {
             use_len = sizeof(int);
+        }
 
         rnd = rand();
-        memcpy( output, &rnd, use_len );
+        memcpy(output, &rnd, use_len);
         output += use_len;
         len -= use_len;
     }
 
-    return( 0 );
+    return 0;
 }
 
-/*
- * Clear some memory that was used to prepare the context
- */
-#if defined(MBEDTLS_ECP_C)
-void ecp_clear_precomputed( mbedtls_ecp_group *grp )
-{
-    if( grp->T != NULL )
-    {
-        size_t i;
-        for( i = 0; i < grp->T_size; i++ )
-            mbedtls_ecp_point_free( &grp->T[i] );
-        mbedtls_free( grp->T );
-    }
-    grp->T = NULL;
-    grp->T_size = 0;
-}
-#else
-#define ecp_clear_precomputed( g )
-#endif
+MBED_NOINLINE static int benchmark_arc4() {
```

Comment:
Why "NOINLINE"? If you allow the compiler to inline this, and it ends up as an empty function (by configuration choices), there shouldn't be any effect (no stub function made that does nothing).

Reply:
The NOINLINE is because without it the compiler will inline functions into main() with high optimization levels and dramatically increase the stack size causing the application to fail...

Reply:
In a later commit I changed location of the preprocessor checks to surround the entire function and the call in main, so there will not be an empty function added when the feature is disabled.

Note that @RonEld had a similar comment, so I changed this. As I discussed with him, I prefer the other approach because it decreases the number of preprocessor checks with only negligible effect in code size and runtime. Please find below my answer to @RonEld's original comment:

>I think code size here is not really a problem. When the feature is not defined as a macro then the compiler (GCC) will generate the following code for an empty function:
```
0000181c <benchmark_havege()>:
    181c:	2000      	movs	r0, #0
    181e:	4770      	bx	lr
```
>Then there will be the following call in the main:
```
    36aa:	f7fe f8b7 	bl	181c <benchmark_havege()>
    36ae:	b9b8      	cbnz	r0, 36e0 <main+0x37c>
```
>Also, I checked and when multiple functions are empty, then the compiler will reuse one of the empty functions instead of emitting multiple ones to reduce code size. In any case, the overhead is 2.5 words (10 bytes) which take about 8-10 clock cycles to execute in a Cortex-M0...